src/nomad_simulations/schema_packages/data_types.py

```diff
@@ -0,0 +1,316 @@
+
+    def convertible_from(self, other):
+        """Check if this data type can convert from another type."""
+        # Follow the same convertibility rules as the base dtype
+        if self._dtype in {int, np.int64}:
+            return other in (int, np.int64, np.int32, np.int16, np.int8)
+        elif self._dtype is np.int32:
+            return other in (np.int32, np.int16, np.int8)
+        elif self._dtype is np.int16:
+            return other in (np.int16, np.int8)
+        elif self._dtype is np.int8:
+            return other is np.int8
+        else:
+            return False
+
+    def serialize_self(self):
+        """Serialize the datatype configuration."""
+        return {
+            'type_kind': 'custom',
+            'type_data': f'{self.__class__.__module__}.{self.__class__.__name__}',
+            'type_bound': str(self.bound),
+        } | self.flags
+
+    def normalize_flags(self, flags: dict):
+        """Reconstruct from serialized data."""
+        bounds_str = flags.get('type_bound', '')
+        self.bound = Bound(bounds_str)
+        # Apply any flags to base datatype
+        if hasattr(super(), 'normalize_flags'):
+            super().normalize_flags(flags)
+        return self
+
+    def normalize(self, value, **kwargs):
+        """Normalize value and validate bounds."""
+        normalized_value = super().normalize(value, **kwargs)
+        return self.bound.check(normalized_value, **kwargs)
+
+    def standard_type(self):
+        """Return the equivalent python type for indexing."""
+        return 'int'
+
+
+class m_float_bounded(InexactNumber):
+    """
+    Bounded float data type.
+
+    Example:
+        m_float_bounded(dtype=float, bound=Bound('[0.0,1.0]'))    # 0.0 ≤ x ≤ 1.0 (floats)
+    """
+
+    __slots__ = ('bound',)
+
+    def __init__(self, dtype=float, bound=None):
+        """Initialize bounded float with dtype and bounds.
+
+        Args:
+            dtype: Float data type, mostly used to specify framework and accuracy (float, np.float64, etc.)
+            bound: Bound instance specifying the valid range
+        """
+        super().__init__(dtype)
+        self.bound = bound or Bound()
+
+    def convertible_from(self, other):
+        """Check if this data type can convert from another type."""
+        # Follow the same convertibility rules as the base dtype
+        if self._dtype in {float, np.float64}:
+            return other in (float, np.float64, np.float32, np.float16)
+        elif self._dtype is np.float32:
+            return other in (np.float32, np.float16)
+        elif self._dtype is np.float16:
+            return other is np.float16
+        else:
+            return False
+
+    def serialize_self(self):
+        """Serialize the datatype configuration."""
+        return {
+            'type_kind': 'custom',
+            'type_data': f'{self.__class__.__module__}.{self.__class__.__name__}',
+            'type_bound': str(self.bound),
+        } | self.flags
```

Comment:
```suggestion
        return return super().serialize_self() | {'type_bound': self.bound}
```

Reply:
So the proper `type_kind` is `python` instead of `custom`? Will amend the tests to reflect this.